### ISSUE TITLE
Glitched/Corrupted Steam UI elements in the headset

### ISSUE THREAD

**Issue by vibudh007** [alvr-org/ALVR#2373]
## Description
SteamVR has glitched/corrupted UI elements in the Quest 2. Using the latest ALVR binary on the PC and ALVR installed on the Quest 2 via Quest store.
Everything else works fine.

![Screenshot from 2024-08-06 16-10-44](https://github.com/user-attachments/assets/52738e98-e763-4fb7-a20c-dfab44e62825)

## General Troubleshooting
- [x] I carefully followed the instructions in the [README](https://github.com/alvr-org/ALVR/blob/master/README.md) and successfully completed the setup wizard
- [x] I read the [ALVR GitHub Wiki](https://github.com/alvr-org/ALVR/wiki)

### Hardware
**Note**: https://linux-hardware.org/?probe=fd9d1ea7b8

**CPU**: AMD Ryzen 9 5900HX 

**GPU**: AMD Radeon 6800M

**GPU Driver Version**: mesa-vulkan-drivers 24.1.5-2.fc40, mesa-va-drivers-freeworld 24.1.5-1.fc40

**Audio**: Pipewire

### Installation
**ALVR Version**: 20.10.0

**ALVR Settings File**: All the settings are at default.

**SteamVR Version**: 2.7

**Install Type**:
- [ ] Packaged (`exe`, `deb`, `rpm`, etc)
- [x] Portable (`zip`)
- [ ] Source

**OS Name and Version**: Fedora Linux 40 (Workstation Edition)

**Comment by emwjacobson** [comment]
Im getting very similar corruption with a similar setup.

https://linux-hardware.org/?probe=1f86e6207d

ALVR Version: 20.11.1
ALVR Settings: All Default
SteamVR Launch Options: `WAYLAND_DISPLAY='' ~/.local/share/Steam/steamapps/common/SteamVR/bin/vrmonitor.sh %command%`

**Comment by InfinityJammyyy** [comment]
Same here issue here corrupt UI, did anyone find a fix for this?

Running alvr on manjaro kde plasma, here is my GPU if anyone can manage to correlate this problem:
AMD ATI Radeon 540X/550X/630 / RX 640 / E9171 MCM

**Comment by InfinityJammyyy** [comment]
Finally fixed the issue, it is as simple as adding "DRI_PRIME=1" to the start of your launch options on SteamVR, SteamVR apparently needs to be run by your dedicated GPU and ALVR seems to not launch it with that in mind (even tho when you click launch SteamVR it says "DRI_PRIME=1 %command%")

If you do not know how to put launch options in a steam game follow this: https://help.steampowered.com/en/faqs/view/7D01-D2DD-D75E-2955

**Comment by emwjacobson** [comment]
Unfortunately adding `DRI_PRIME=1` is not a universal fix, not for me at least.

I used `DRI_PRIME=1 ~/.local/share/Steam/steamapps/common/SteamVR/bin/vrmonitor.sh %command%`, yet still experience corruption in menus and am unable to launch games (Stuck on "Up Next" screen). Not sure if it has to do with the fact that my machine has both an iGPU and dGPU, either.

**Comment by MangoTCF** [comment]
I am experiencing the same using my Framework 16(AMD Radeon 780M/AMD Radeon RX 7700S hybrid) with Arch. Although in my experience, steam cannot launch with DRI_PRIME=1 at all, and launching only SteamVR with the option results in the glitches.

**Comment by JustinSolstice** [comment]
Same thing is happening here too, using Arch and wayland with amd gpu

**Comment by The-personified-devil** [comment]
See [this section on the wiki](https://github.com/alvr-org/ALVR/wiki/Linux-Troubleshooting#steamvr-dashboard-not-rendering-in-vr-on-nvidia-discrete-gpu)

**Comment by JustinSolstice** [comment]
> See [this section on the wiki](https://github.com/alvr-org/ALVR/wiki/Linux-Troubleshooting#steamvr-dashboard-not-rendering-in-vr-on-nvidia-discrete-gpu)

This only works for nvidia GPUs, sorry for necroing the issue but am still having issues with it

**Comment by Wgelyjr** [comment]
Fedora 43, same issue. No amount of DRI_PRIME=1 seems to help... 

Odd that it only happens in game. Changing any encoding settings does not seem to affect the problem.

**Comment by RSDNTWK** [comment]
We've experienced the same issues as @MangoTCF using a Pico 4 regarding this with multiple distros. In the base Arch, Fedora, Debian and Ubuntu  distros, we've experienced crashes or both the UI and Games being broken entirely even when attempting to use the wayland and DRI_PRIME commands to attempt to fix them. With Kubuntu the UI works without needing the wayland fix however the DRI_PRIME command has no effect trying to fix the game graphics. Bazzite used to work at some point in version 41 with both the wayland and DRI_PRIME commands applied but recently broke in an update within the last 3 months and runs into the same issues as the other distros mentioned.

**Comment by Nocrex** [comment]
I was also experiencing this issue, even after setting DRI_PRIME=1 in the SteamVR and the game's launch option.

But I was now able to find a fix for it, by running steam with `DRI_PRIME=<vendor_id>:<device_id>`
To get the vendor and device id of the correct gpu you can use `MESA_VK_DEVICE_SELECT=list vulkaninfo`, it's the hexadecimal numbers with a colon inbetween right after "GPU x:".

For me if I set this variable only in the SteamVR launch options, the dashboard stays corrupted, but games start working correctly. To make the dashboard display correctly I need to launch steam with the variable set.

Also there is an issue where steam goes into a loop while starting with this variable set, and to get around that I had to disable gpu acceleration in the steam settings.

**Comment by RSDNTWK** [comment]
This was not an issue on an older version of bazzite which may have been 39 or 40 as we were able to use `DRI_PRIME=1 QT_QPA_PLATFORM=xcb ~/.local/share/Steam/steamapps/common/SteamVR/bin/vrmonitor.sh %command%` which resolved both the UI corruption and the rendering issues for games like Waltz of the Wizard. We suspect a bug that was fixed in Steam, SteamVR or Mesa suddenly broke again during recent testing. It's doubtful that ALVR itself is the root cause as the issues were present on the SteamVR display view. More extensive testing would be needed to debug it to see which is the culprit.

**Comment by RSDNTWK** [comment]
We've found the root cause. It seems that the steam client itself is preventing both the DRI_PRIME and QT_QPA_PLATFORM=xcb commands in the steam launch options from running properly. The solution we discovered is to run SteamVR directly by executing the  vrmonitor.sh script using a terminal instead of the alvr launch or Steam itself. We used the following command in the terminal to launch SteamVR: `DRI_PRIME=1 QT_QPA_PLATFORM=xcb ~/.local/share/Steam/steamapps/common/SteamVR/bin/vrmonitor.sh` (you may need to change the path of your Steam install.) We can confirm all vr games including ones using proton launch without any extra commands. An external script to change how ALVR's "Launch SteamVR" option launches would be recommended instead of calling SteamVR within Steam itself. We hope this helps everyone involved.

**Comment by Marhau-devoloper** [comment]
may i found a fix. I tried alvr on arch, monjaro Ubuntu, Linux mint and it doesn't work on any distro than i try to experimentation with launching options and it work if use DRI_PRIME=1 in start it did not work i just remove it and leave 
_WAYLAND_DISPLAY='' /home/YOURUSERNAME/.steam/debian-installation/steamapps/common/SteamVR/bin/vrmonitor.sh %command%_

**Comment by silbinarywolf** [comment]
If anyone is discovering this in the future and has an AMD graphics card, using `DRI_PRIME=0` instead of ` DRI_PRIME=1` might be the fix for you!

Documentation for ALVR *assumes* that DRI_PRIME=1 will definitely be your dedicated graphics card but for me and the reddit user at this link here, that was not the case: https://www.reddit.com/r/linux_gaming/comments/wjkqtw/dri_prime_operates_backwards/

My integrated Graphics card:
```sh
DRI_PRIME=1 glxinfo | grep "OpenGL renderer"
OpenGL renderer string: AMD Ryzen 9 9950X3D 16-Core Processor (radeonsi, raphael_mendocino, LLVM 21.1.8, DRM 3.64, 6.17.7-ba25.fc43.x86_64)
```

My dedicated graphics card:
```sh
DRI_PRIME=0 glxinfo | grep "OpenGL renderer"
OpenGL renderer string: AMD Radeon RX 9070 XT (radeonsi, gfx1201, LLVM 21.1.8, DRM 3.64, 6.17.7-ba25.fc43.x86_64)
```

**Comment by silbinarywolf** [comment]
I have raised a separate issue here related to the `DRI_PRIME` issue here: https://github.com/alvr-org/ALVR/issues/3201

**Comment by vibudh007** [comment]
Hey everyone, I'll add another fix to this thread.

After a few months I dusted off my Quest 2 and revisited VR games again, and I was hit with all the same issues. I tried both Steam Link and ALVR, and both have their pros and cons. Steam Link now works (yay!!) and works very well I must say, but barely gives you any control over the data stream, unlike ALVR.

Anyway, the first problem I had was glitching/corruption of the entire video (completely unusable VR). This was easy enough to solve with having
`DRI_PRIME=1 ~/.local/share/Steam/steamapps/common/SteamVR/bin/vrmonitor.sh %command%`
in the launch options of SteamVR.
I could now get into SteamVR and even play games, but the issue (THE issue, the one this thread is about) came up that the dashboard won't render properly.

I believe I understand why this happens in the first place. The SteamVR dashboard is actually webview, and that webview launches on a different thread and doesn't respect the DRI_PRIME=1 environment variable, making the dashboard run on the dGPU and messing up the decoding of the data stream, or at least I think that's what happens. BTW it's called vrwebhelper, it's in the same parent directory as vrmonitor.sh of launch options fame.

To make vrwebhelper launch with the same DRI_PRIME=1 environment variable, one can launch steam itself with DRI_PRIME=1. Doing this from the terminal however, lead me to another famous issue, the Steam Client window not being able to render at all. Steam is up and running, but the Steam Client has a launch parameter 'PrefersNonDefaultGPU' which is by default set to true, and running DRI_PRIME=1 along with this parameter confuses the Steam Client.
To fix this, one has to change the launch parameter to false. 

The easy way to do this is to modify the ~Desktop/steam.desktop entry and change the parameter to false, but this results in SteamVR only working properly when launched from the desktop. A better way is to modify /usr/share/applications/steam.desktop and change PrefersNonDefaultGPU to false, and prepend env DRI_PRIME=1 to Exec. For Bazzite and other immutable distros, one can copy the steam.desktop file into the ~/.local/share/applications/ directory using
`cp /usr/share/applications/steam.desktop ~/.local/share/applications` 
and modifying ~/.local/share/applications/steam.desktop and change PrefersNonDefaultGPU to false and prepend env DRI_PRIME=1 to Exec. Then to reload the desktop entries, run
`update-desktop-database /usr/share/applications/steam.desktop`
And now Dashboard works perfectly, as it's spawned as a subthread under Steam, whose environment variable dictates all subthreads use the dGPU.

Tl;dr:
1. `cp /usr/share/applications/steam.desktop ~/.local/share/applications` 
2. Edit `~/.local/share/applications/steam.desktop`
3. Change `PrefersNonDefaultGPU=true` to `PrefersNonDefaultGPU=false`
4. Change `Exec=/usr/bin/steam` to `Exec=env DRI_PRIME=1 /usr/bin/steam`
5. Save and exit the editor.
6. Add `DRI_PRIME=1 ~/.local/share/Steam/steamapps/common/SteamVR/bin/vrmonitor.sh %command%` to the launch options of SteamVR
7. Add `DRI_PRIME=1 %command%` to the launch options of every VR game you wanna play.
8. ...
9. Profit!

**Comment by RSDNTWK** [comment]
@vibudh007 We did attempt this ourselves many times however due to our dual GPU setup the only solution we discovered was to completely disable the iGPU at bootup as setting DRI_PRiME=1 on steam itself caused it to fail to start altogether.

**Comment by vibudh007** [comment]
@RSDNTWK By 'fail to start' you wouldn't happen to mean in a similar way to what I described. right? The Steam window opens and then immediately closes once every few seconds, infinitely 'bootlooping'?? If so, PrefersNonDefaultGPU=true was the fix for me.

But if not, I guess it also depends on the multi-GPU setup itself. Maybe its different for tower PCs versus laptops, or between the many configurations of AMD/Intel CPU/iGPU + AMD/NVidia dGPU. In which case, I dunno, maybe capitalism is to blame.

At least for my ROG Strix G15 Advantage Edition with Ryzen 9 5900HX + Radeon 6800M, the steps I mentioned fixed my SteamVR.
